engine/src/main/scala/cromwell/engine/workflow/WorkflowDockerLookupActor.scala

```diff
@@ -0,0 +1,322 @@
+package cromwell.engine.workflow
+
+import akka.actor.{ActorRef, LoggingFSM, Props}
+import cromwell.core.{Dispatcher, WorkflowId}
+import cromwell.database.sql.SqlDatabase
+import cromwell.database.sql.tables.DockerHashStoreEntry
+import cromwell.docker.DockerHashActor.{DockerHashFailureResponse, DockerHashSuccessResponse}
+import cromwell.docker.{DockerClientHelper, DockerHashRequest, DockerHashResult, DockerImageIdentifier}
+import cromwell.engine.workflow.WorkflowActor.{RestartExistingWorkflow, StartMode}
+import cromwell.engine.workflow.WorkflowDockerLookupActor._
+import cromwell.services.SingletonServicesStore
+import lenthall.util.TryUtil
+
+import scala.concurrent.duration._
+import scala.language.postfixOps
+import scala.util.{Failure, Success, Try}
+
+/**
+  * Ensures docker hash consistency throughout a workflow.
+  *
+  * Caches successful docker hash lookups and serve them to subsequent identical requests.
+  * Persists those hashes in the database to be resilient to server restarts.
+  *
+  * Failure modes:
+  * 1) Failure to load hashes from the DB upon restart.
+  * 2) Failure to parse hashes from the DB upon restart.
+  * 3) Failure to write a hash result to the DB.
+  * 4) Failure to look up a docker hash.
+  * 5) Timeout from DockerHashActor.
+  *
+  * Behavior:
+  * 1-2) Return a terminal lookup failure for all pending requests, transition to a permanently Failed state in which any
+  *      future requests will immediately return lookup failure.  The JobPreparation actor should fail in response to this
+  *      lookup termination message, which in turn should fail the workflow.
+  * 3-4) Return lookup failure for the current request and all pending requests for the same tag.  Any future requests
+  *      for this tag will be attempted again.
+  * 5)   Return lookup failure for all pending requests.  Any future requests for these tags will be attempted again.
+  */
+
+class WorkflowDockerLookupActor private[workflow](workflowId: WorkflowId, val dockerHashingActor: ActorRef, startMode: StartMode, databaseInterface: SqlDatabase)
+  extends LoggingFSM[WorkflowDockerLookupActorState, WorkflowDockerLookupActorData] with DockerClientHelper with SingletonServicesStore {
+
+  implicit val ec = context.system.dispatchers.lookup(Dispatcher.EngineDispatcher)
+
+  // Amount of time after which the docker request should be considered lost and sent again.
+  override protected def backpressureTimeout: FiniteDuration = 10 seconds
+  // Amount of time to wait when we get a Backpressure response before sending the request again.
+  override protected def backpressureRandomizerFactor: Double = 0.5D
+
+  context.become(dockerReceive orElse receive)
+
+  {
+    val restart = startMode == RestartExistingWorkflow
+    val initialState = if (restart) AwaitingFirstRequestOnRestart else Running
+
+    startWith(initialState, WorkflowDockerLookupActorData.empty)
+  }
```

Comment:
Preventing pollution of this Actor's surface area with variables only used in this narrow scope.  But if the tradeoff is confusion for future maintainers, maybe it's not worth it.  😄